### ISSUE TITLE
[Feature Request] Ability to check whether the expectations are met and reset mock expectations

### ISSUE THREAD

**Issue by hardikmodha** [golang/mock#459]
### Requested feature: 
- **Ability to check whether all expectations are met** 
Currently, We only have `EXPECT()` which we can use in the user code and the rest (i.e. checking whether all expected calls were made) happens inside the controller. I would like to propose a utility which allows user code to check whether all expectations are met or not. A feature similar to the one which [go-sqlmock](https://godoc.org/github.com/DATA-DOG/go-sqlmock#SqlmockCommon) provides, namely `ExpectationsWereMet`
- **Resetting the mock behavior so that any previous behavior in not propagated to other tests** 
Currently, We need to redefine the `controller` to define the new mock behavior. I would like to propose a utility which allows user code to reset all previous expectations so that test separation is maintained and dirty state of other tests is not propagated to other tests. 
 
**Why the feature is needed** 
Right now, In my current project, While doing component-level testing, I need to inject the mocked dependencies of other services when my component gets initialized. So the controller is also defined there only. All these dependencies gets called in the code. It's time and resource consuming to initialize the application on each and every test. So if the above features are added then it would really help. 

<details>
<summary>The following is the high-level code structure on how I'm planning on consuming these features
</summary>

For the sake of brevity I've included only high-level part.
```go
file: ct_main_test.go // This defines the TestMain.go and initializes the application

var remoteService

func TestMain(m *testing.M) {
    // initialize application
    ctrl := controller.NewController(helper)
    remoteService = mock.MockedRemote(ctrl)
    service = initializeMyService(remoteService)    
    .
    .
    os.Exit(m.Run())
}

file: ct_some_file_test.go

func Test_xxx(t *testing.T) {
      // Before proceeding, Reset any previous expectations
      remoteservice.ResetExpectations()

      // Given
      remoteService.Expect().ServiceMethod().Times(2)

     // When an API is called
     myService.SayHello()

     // Then
     remoteService.ExpectationsWereMet()    // Assert that all expectations were met. If expectations were not met, we throw an error
}
```
</details>

This would really help in implementing best practices in my current project and will help in avoid passing the dirty state between tests mistakenly.

**PS**: Excuse me if it's possible somehow using the current features provided by `gomock`. In that case, Let me know how can I achieve the same. Thanks!

**Comment by cvgw** [comment]
Making sure I understand this request;

Your tests require a lot of set up, particularly in the mocking of expectation?
Your tests re-use a lot of the same setup?
You want to use the same mocked expectation between tests by "resetting" (clearing the received calls) the mocks so that the expectations don't have to be set up again?

Apologies if I've misunderstood any part of this.

I suppose my initial question is why it is not sufficient to put the mocks/setup into a help function within your test code?

something like

```
func setUpMocks(ctrl *gomock.Controller) (mock1, mock2, mock3, error) {
  // set up the mocks
  return mock1, mock2, mock3, nil
}

func TestFooBar(t *testing.T) {
  ctrl := gomock.NewController(t)
  mock1, mock2, mock3, err := setUpMocks(ctrl)
  // run tests
}
```

**Comment by hardikmodha** [comment]
Thank you @cvgw for the response. 

>You want to use the same mocked expectation between tests by "resetting" (clearing the received calls) the mocks so that the expectations don't have to be set up again?

No, I want to reset the expectations + received calls as well. Not sure if it's possible but we also need to make sure that test pollution doesn't happen.

Something like
```
mock1, mock2, mock3, err := setUpMocks(ctrl)

func Test_xxx(t *testing.T) {
      // Before proceeding, Reset any previous expectations and received calls
      mock1.ResetExpectationsAndReceivedCalls()

      // Given
      mock1.Expect().ServiceMethod().Times(2)

     // When an API is called
     myService.SayHello()

     // Then
     mock1.ExpectationsWereMet()    // Assert that all expectations were met. If expectations were not met, we throw an error
}
```
and likewise. 

Regarding your suggested approach, I can very well do it but the test is a module-level test that sets up the database and runs the application. Calls internal to the module are not mocked. Only the calls for other 
(dependent) modules are mocked. (so it's sort of an integration test but at module level only - not at the whole system level) 

In the example you provided, mocks returned from `setUpMocks` needs to be injected when my application starts, so I will have to start the application for each and every test. 

Something like, 
```
mock1, mock2, mock3, err := setUpMocks(ctrl)

initializeMyApplication(mock1, mock2, mock3)
```
If I don't inject the newly setup mocks in my application, then calls to external modules will fail. 

In short, I create a mock at a global level by passing a controller, and then before/after each and every test I should be able to reset the mock calls + expectations so that I can use the same mock instance. I hope I'm making sense :sweat_smile:

**Comment by tomarrell** [comment]
I would second this. This would be useful to cut down on the setup code between tests which are run as subtests. A pattern which would look something like:

```go

TestSomething(t *testing.T) {
    // setup mocks
    // dependencies with injected mocks

    t.Run("subtest 1", func(t *testing.T) {
        // setup mock expectations
        // run test code
        // assert mock expections and reset
    })

     t.Run("subtest 2", func(t *testing.T) {
        // setup mock expectations
        // run test code
        // assert mock expections and reset
    })

    // ...
}

```

**Comment by cvgw** [comment]
@hardikmodha Thanks for the additional clarification. I think I understand your use-case now and why this would be helpful. 

Will think on it and chat with the other maintainers.

**Comment by hardikmodha** [comment]
@cvgw @codyoss Thinking out loud, How about we introduce a new construct `Continue` on the `controller`. It will be almost like `ctrl.Finish` with the following differences.

1. It can be called multiple times from the user code.
2. It will implicitly reset the calls at the end.

Above will ensure that when the user code calls `ctrl.Continue()` all the expectations are checked for and calls are also reset. Now, user code can reuse the same controller to define new expectations. Older expectations are no longer in the picture. This way we can be sure that unmet expectations of any tests don't have side-effects on other tests.

**Comment by hardikmodha** [comment]
@tomarrell What do you think of the proposed solution? In the example you mentioned, We can use `ctrl.Continue()` as follows.

```
TestSomething(t *testing.T) {
    ctrl := gomock.NewController(t) // setup mocks
    mockedInterface := setupDependencies(ctrl) // dependencies with injected mocks 

    t.Run("subtest 1", func(t *testing.T) {
        mockedInterface.EXPECT().method() // setup mock expectations
        defer ctrl.Continue() // assert that all expectations are met and reset
        // run test code
    })

     t.Run("subtest 2", func(t *testing.T) {
        mockedInterface.EXPECT().method().Return(1).Times(1) // setup mock expectations
        defer ctrl.Continue() // assert that all expectations are met and reset
        // run test code
    })

    // ...
}
```

**Comment by nekinie** [comment]
Do we need to do anything around passing the sub tests testing.T in?

**Comment by hardikmodha** [comment]
Sorry, Didn't get your comment @nekinie. Can you please elaborate?

**Comment by nekinie** [comment]
In the above solution, unexpected calls to a mock function within subtest 1 would trigger a failure on "TestSomething" instead of "TestSomething/subtest _1". Could we pass subtest 1's *testing.T into defer ctrl.Continue() so the failures are reported for the correct test?

**Comment by cvgw** [comment]
@hardikmodha My understanding (and this may be imperfect as I haven't been involved with the project since it's inception) is that the original design and intention of gomock is to be used within the scope of a single test or subtest. 

I don't know every place that this idea is baked into the code, but one example would be the creation of a `Controller` instance using `*testing.T`. If `Controller` is created using a parent test (IE `TestFoo`) what happens when a subtest fails and `t.Fail()` is called on the parent `*testing.T` rather than the subtest `*testing.T`? I would guess probably not the behavior that users would want.

(IE)

```
func TestFoo(t *testing.T) {
  ctrl := NewController(t)
  mockFoo := NewMockFoo(ctrl)

  t.Run("subtest 1", func(t *testing.T) {
    // do something with mockFoo which causes a missed mock
  })
}
```

Reusing mocks across tests is not the direction that the current maintainers think the project should go, but it isn't completely out of the question. For me at least, as a start I would want to see 100% of the current test suite ported over to test using a `Controller` created outside of the current test context. Though there would probably need to be some discussion about how to reliable test that state.

**Comment by codyoss** [comment]
For subtests if you do not create a new Controller with the subtest t you will get errors from the `testing` package.

Example:
`gm459.go`
```go
package gm459

//go:generate mockgen -source=gm459.go -destination=gm459_mock.go -package=gm459

type Foo interface {
	Bar() string
}

type ReallyExpensiveService struct {
	f Foo
}

func NewReallyExpensiveService(f Foo) *ReallyExpensiveService {
	return &ReallyExpensiveService{f}
}

func (s *ReallyExpensiveService) Baz() string {
	return s.f.Bar()
}
```

`gm459_test.go`
```go
package gm459

import (
	"testing"

	"github.com/golang/mock/gomock"
)

func TestBaz(t *testing.T) {
	ctrl := gomock.NewController(t)
	mockFoo := NewMockFoo(ctrl)
	svc := NewReallyExpensiveService(mockFoo)

	t.Run("one", func(t *testing.T) {
		//mockFoo.EXPECT().Bar().Return("one")
		svc.Baz()
	})

        // Will never get called
	t.Run("two", func(t *testing.T) {
		mockFoo.EXPECT().Bar().Return("one")
		svc.Baz()
	})
}
```

Output:

```bash
$ go test -v
=== RUN   TestBaz
=== RUN   TestBaz/one
=== CONT  TestBaz
    gm459.go:18: Unexpected call to *gm459.MockFoo.Bar([]) at /tmp/gm459.go:18 because: there are no expected calls of the method "Bar" for that receiver
=== CONT  TestBaz/one
    testing.go:1033: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
--- FAIL: TestBaz (0.00s)
    --- FAIL: TestBaz/one (0.00s)
FAIL
exit status 1
```

**Comment by hardikmodha** [comment]
Apologies for getting back late. I would circle back to this later today.

**Comment by hardikmodha** [comment]
Hmmm, I got your concern. I tried passing the `t *testing.T` to `Continue` as @nekinie mentioned earlier but that only helped to avoid error from the `testing` package in `Continue` function. But we have `ctrl.T.Errorf` and `ctrl.T.Fatalf` at other places which don't have reference to `t` of subtests and thus `test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test`

**Comment by hardikmodha** [comment]
I ended up doing a separate setup and teardown for all the tests and making them run in parallel.
